### PR TITLE
module: add :os exwm

### PR DESCRIPTION
**Pull request by LemonBreezes**
Hello everyone, I have written an EXWM module for Doom Emacs, which I hope you will all enjoy. 

Some features of note from this module are:
- Evil support for all EXWM buffers, with the entire suite of commands supported on Firefox-based and Chrome-based web browsers.
- The ability to edit any input field using Emacs with language detection for selecting the appropriate major mode. 

I was going to include a feature for automatically creating/destroying workspaces for EXWM applications but decided not to include it in the initial pull request. I would love to include it in a module flag.

### CODE REVIEW

File: modules/os/exwm/README.org [hunk 1]
```diff
@@ -0,0 +1,145 @@
+#+TITLE:   private/exwm
```
Comment (jeetelongname) [inline modules/os/exwm/README.org:1]: Should this not be `os/exwm`?

File: modules/os/exwm/README.org [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
++ @[[https://github.com/LemonBreezes][LemonBreezes]] (Author)
+
+** Module Flags
+
+This module provides no flags.
+
+** Plugins
+
+# A list of linked plugins
```
Comment (jeetelongname) [inline modules/os/exwm/README.org:45]: This should probably be removed

File: modules/os/exwm/README.org [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+** Setting up your ~xinitrc~ file.
+
+In order to enable EXWM, Emacs needs to be running within X.Org without any
+window manager present. The recommended way to go about this is with the
+~startx~ script. The [[https://github.com/ch11ng/exwm/wiki/Configuration-Example][EXWM wiki]] contains an example ~xinitrc~ file for use by
+~startx~.
+
+** Setting the default major mode for EXWM Edit.
+:PROPERTIES:
```
Comment (jeetelongname) [inline modules/os/exwm/README.org:90]: Don't think this property tag is meant to be here

File: modules/os/exwm/README.org [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+:END:
+
+For users of Evil, the ~f~ and ~F~ keys are bound to special Link Hint commands
+within Chrome-based and Firefox-based web browsers. For these commands to work,
+you must have the Link Hint add-on/extension installed. For Chrome-based web
+browsers, install Link Hint through [[https://chrome.google.com/webstore/detail/link-hints/kjjgifdfplpegljdfnpmbjmkngdilmkd][here]]. For Firefox-based web browsers, use
+[[https://addons.mozilla.org/en-US/firefox/addon/linkhints/?utm_source=addons.mozilla.org&utm_medium=referral&utm_content=search][this]] link instead.
+
+** Setting up multiple monitors
+:PROPERTIES:
```
Comment (jeetelongname) [inline modules/os/exwm/README.org:113]: Same here

File: modules/os/exwm/README.org [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+:CREATED_TIME: [2022-01-06 Thu 00:06]
+:END:
+
+Currently this module does not provide any simplified setup for multi-head
+configurations. Please refer to
+https://github.com/ch11ng/exwm/wiki#randr-multi-screen for how to do this.
+
+* Troubleshooting
+
+# Common issues and their solution, or places to look for help.
```
Comment (jeetelongname) [inline modules/os/exwm/README.org:123]: Remove this

File: modules/os/exwm/config.el [hunk 6]
```diff
@@ -0,0 +1,116 @@
+;;; private/exwm/config.el -*- lexical-binding: t; -*-
+
+(add-hook 'exwm-update-title-hook #'+exwm-rename-buffer-to-title)
+
+(use-package! exwm
```
Comment (jeetelongname) [inline modules/os/exwm/config.el:5]: You should not nest use-package! calls, instead use the `:after` keyword, they are functionally the same but the second is cleaner and more importantly doom convention

File: modules/os/exwm/packages.el [hunk 7]
```diff
@@ -0,0 +1,12 @@
+;; -*- no-byte-compile: t; -*-
+;;; private/exwm/packages.el
+
+(package! exwm :pin "10bd12234e896d35a2c4eafabc62a31126d23bf3")
+(when (featurep! :editor evil)
```
Comment (jeetelongname) [inline modules/os/exwm/packages.el:5]: Should probably be `evil +everywhere`

File: modules/os/exwm/packages.el [hunk 8]
```diff
@@ -0,0 +1,12 @@
+;; -*- no-byte-compile: t; -*-
+;;; private/exwm/packages.el
```
Comment (jeetelongname) [inline modules/os/exwm/packages.el:2]: again change to `os/exwm`

File: modules/os/exwm/config.el [hunk 9]
```diff
@@ -0,0 +1,116 @@
+;;; private/exwm/config.el -*- lexical-binding: t; -*-
```
Comment (jeetelongname) [inline modules/os/exwm/config.el:1]: Change to `os/exwm`

File: modules/os/exwm/autoload/exwm.el [hunk 10]
```diff
@@ -0,0 +1,66 @@
+;;; private/exwm/autoload/exwm.el -*- lexical-binding: t; -*-
```
Comment (jeetelongname) [inline modules/os/exwm/autoload/exwm.el:1]: change to `os/exwm`

File: modules/os/exwm/autoload/exwm-firefox.el [hunk 11]
```diff
@@ -0,0 +1,15 @@
+;;; private/exwm/autoload/exwm-firefox.el -*- lexical-binding: t; -*-
```
Comment (jeetelongname) [inline modules/os/exwm/autoload/exwm-firefox.el:1]: change to `os/exwm`

File: modules/os/exwm/autoload/exwm-edit.el [hunk 12]
```diff
@@ -0,0 +1,69 @@
+;;; private/exwm/autoload/exwm-edit.el -*- lexical-binding: t; -*-
```
Comment (jeetelongname) [inline modules/os/exwm/autoload/exwm-edit.el:1]: change to `os/exwm`

**Review by jeetelongname — CHANGES_REQUESTED** [review]
Hi! Thanks for your work, as mentioned "reasonable defaults" are up in the air for this module so it may be that this module does not get into doom core, that being said there is also the contrib repo coming for modules that are community maintained so this work will defo not go to waste.

anyway I have reviewed the module in so far and found some small things that need changing.

thanks for your work!

**Comment by jeetelongname** [comment]
also can you please rebase your branch to conform to our [git conventions](https://gist.github.com/hlissner/4d78e396acb897d9b2d8be07a103a854) I would also split this out into many commits detailing individual changes

**Comment by LemonBreezes** [comment]
Thank you. I have just implemented the fixes you stated. I will work on splitting this into several commits following the Git conventions.

**Comment by jeetelongname** [comment]
You don't need to close the PR you should be able to turn it into a draft which keeps it open but makes it clear its incomplete (it can't be merged until you put it up for review)

**Comment by LemonBreezes** [comment]
> You don't need to close the PR you should be able to turn it into a draft which keeps it open but makes it clear its incomplete (it can't be merged until you put it up for review)

Ahh, thank you. I finally found that button. I am not very familiar with Git so it's going to take me a bit of time to figure this out.

**Comment by maikol-solis** [comment]
Hi @LemonBreezes   

I work mostly with R projects, and I'm starting using [ESS](https://github.com/hlissner/doom-emacs/tree/master/modules/lang/ess). 

I would be super nice to have plots embedded inside doom-emacs using exwm. I just stumble upon with [this post ](https://emacs.stackexchange.com/a/42479/21501) and I'm wondering if this fits in your module. WDYT?

Thanks.

**Comment by LemonBreezes** [comment]
> Hi @LemonBreezes
> 
> I work mostly with R projects, and I'm starting using [ESS](https://github.com/hlissner/doom-emacs/tree/master/modules/lang/ess).
> 
> I would be super nice to have plots embedded inside doom-emacs using exwm. I just stumble upon with [this post ](https://emacs.stackexchange.com/a/42479/21501) and I'm wondering if this fits in your module. WDYT?
> 
> Thanks.

Sure. That seems like an interesting alternative to using temporary files with `auto-revert-mode`. I am definitely interested in having interesting uses of EXWM added to this module. The module is currently living in [exwm-module-for-doom-emacs](https://github.com/LemonBreezes/exwm-module-for-doom-emacs). Soon I will close this pull request and redo this module for the master branch.

**Comment by LemonBreezes** [comment]
So here is my status update: I have been working on this module privately. I plan to release it when Doom gets a public repository for modules. I've dropped support for non-Evil keybindings currently because I am using Evil again. The module is currently here https://github.com/LemonBreezes/exwm-module-for-doom-emacs, though if you want the latest updates, you can go directly to my private config https://github.com/LemonBreezes/.doom.d/tree/master/modules/private/exwm. I regularly have been making major changes so don't expect this module to be stable yet. That being said, this is a really cool project and I use it every day as my daily driver so I work on it regularly.

**Comment by oatmealm** [comment]
Looking forward to this feature added. Thanks!

**Comment by iT-Boyer** [comment]
when run `M-x` or `recenterf file`：
```
Error in post-command-hook (ivy--queue-exhibit): (wrong-type-argument numberp t) 
```

**Comment by LemonBreezes** [comment]
This module is outdated and needs to be rewritten.

**Comment by hlissner** [comment]
As of v2.2, modules have been moved out of core, so module development has moved to [doomemacs/modules](https://github.com/doomemacs/modules) and [doomemacs/modules-contrib](https://github.com/doomemacs/modules-contrib). If you're still interested, the modules-contrib repo is ready to receive PRs like this one, given two small changes:

- The two commits be squashed into one, 
- The first line of the commit message be `module: add :os exwm` (to conform to our git conventions).

If you're not interested, that's fine too. Sorry for the trouble and thanks your for your interest!

**Pull request closed without merging**